### ISSUE TITLE
_FIXED_numpy_to_python errors with precision

### ISSUE THREAD

**Issue by mayfield** [snowflakedb/snowflake-connector-python#9]
When enabling numpy on my connection I run into errors parsing numbers with precision.  They are interpreted as **FIXED** type and passed to `_FIXED_numpy_to_python` as a string value;  This function simply runs `numpy.int64` which will never return a precision based value (eg float/decimal).

In this trace the column is a **NUMBER(38,1)** and the value is `1234.1`.

It appears that **FIXED** types should result in a precision-able type when `precision != 0`.  When numpy is disabled the high precision `Decimal` type is used.  Should this not also be the case in numpy?  I don't see an equivalent in numpy to `Decimal` so perhaps the same logic used without numpy could apply except it would return numpy.int64 when `precision == 0`?  

(Note that even this approach could be considered suspect because **NUMBER** can be 38 digits which far exceeds the int64 type (but not builtin `int`).  One might argue that numpy.int64 is always inadequate for **NUMBER** types.)
```
Command 'main' error, traceback...
Traceback (most recent call last)
2.  File "/usr/lib/python3.5/site-packages/snowflake_connector_python-1.3.14-py3.5.egg/snowflake/connector/cursor.py", line 856, in _row_to_python
      row[idx] = conv_method(col, conv_ctx)
1.  File "/usr/lib/python3.5/site-packages/snowflake_connector_python-1.3.14-py3.5.egg/snowflake/connector/converter.py", line 115, in _FIXED_numpy_to_python
      return numpy.int64(value)
ValueError: invalid literal for int() with base 10: '1234.1'

  During handling of the above exception, another exception occurred:

  Traceback (most recent call last)
  10. File "/home/mayfield/project/shellish/shellish/session.py", line 119, in execute
        raise e
  9.  File "/home/mayfield/project/shellish/shellish/session.py", line 116, in execute
        result = command.run_wrap(args)
  8.  File "/home/mayfield/project/shellish/shellish/command/command.py", line 169, in run_wrap
        raise e
  7.  File "/home/mayfield/project/shellish/shellish/command/command.py", line 165, in run_wrap
        result = self.run(args)
  6.  File "/home/mayfield/project/shellish/shellish/command/autocommand.py", line 52, in run
        return self.func(*positionals, **keywords)
  5.  File "bench_scale.py", line 69, in main
        row = q.fetchone()
  4.  File "/usr/lib/python3.5/site-packages/snowflake_connector_python-1.3.14-py3.5.egg/snowflake/connector/cursor.py", line 689, in fetchone
        return self._row_to_python(row) if row is not None else None
  3.  File "/usr/lib/python3.5/site-packages/snowflake_connector_python-1.3.14-py3.5.egg/snowflake/connector/cursor.py", line 871, in _row_to_python
        u'errno': ER_FAILED_TO_CONVERT_ROW_TO_PYTHON_TYPE,
  2.  File "/usr/lib/python3.5/site-packages/snowflake_connector_python-1.3.14-py3.5.egg/snowflake/connector/errors.py", line 95, in errorhandler_wrapper
        cursor.errorhandler(connection, cursor, errorclass, errorvalue)
  1.  File "/usr/lib/python3.5/site-packages/snowflake_connector_python-1.3.14-py3.5.egg/snowflake/connector/errors.py", line 71, in default_errorhandler
        done_format_msg=errorvalue.get(u'done_format_msg'))
  InterfaceError: 252005: None: Failed to convert: field DEC2: FIXED::1234.1, Error: invalid literal for int() with base 10: '1234.1'
```

**Comment by mayfield** [comment]
Confirmed that even with scale=0 that numpy.int64 is easy to overflow where python's builtin `int` works..

```
Command 'main' error, traceback...
Traceback (most recent call last)
1.  File "/usr/lib/python3.5/site-packages/snowflake_connector_python-1.3.14-py3.5.egg/snowflake/connector/cursor.py", line 860, in _row_to_python
      row[idx] = converter(col)
OverflowError: Python int too large to convert to C long

  During handling of the above exception, another exception occurred:

  Traceback (most recent call last)
  10. File "/home/mayfield/project/shellish/shellish/session.py", line 119, in execute
        raise e
  9.  File "/home/mayfield/project/shellish/shellish/session.py", line 116, in execute
        result = command.run_wrap(args)
  8.  File "/home/mayfield/project/shellish/shellish/command/command.py", line 169, in run_wrap
        raise e
  7.  File "/home/mayfield/project/shellish/shellish/command/command.py", line 165, in run_wrap
        result = self.run(args)
  6.  File "/home/mayfield/project/shellish/shellish/command/autocommand.py", line 52, in run
        return self.func(*positionals, **keywords)
  5.  File "bench.py", line 68, in main
        while q.fetchone() is not None:
  4.  File "/usr/lib/python3.5/site-packages/snowflake_connector_python-1.3.14-py3.5.egg/snowflake/connector/cursor.py", line 689, in fetchone
        return self._row_to_python(row) if row is not None else None
  3.  File "/usr/lib/python3.5/site-packages/snowflake_connector_python-1.3.14-py3.5.egg/snowflake/connector/cursor.py", line 875, in _row_to_python
        u'errno': ER_FAILED_TO_CONVERT_ROW_TO_PYTHON_TYPE,
  2.  File "/usr/lib/python3.5/site-packages/snowflake_connector_python-1.3.14-py3.5.egg/snowflake/connector/errors.py", line 95, in errorhandler_wrapper
        cursor.errorhandler(connection, cursor, errorclass, errorvalue)
  1.  File "/usr/lib/python3.5/site-packages/snowflake_connector_python-1.3.14-py3.5.egg/snowflake/connector/errors.py", line 71, in default_errorhandler
        done_format_msg=errorvalue.get(u'done_format_msg'))
  InterfaceError: 252005: None: Failed to convert: field DEC_NUM: FIXED::29824269732928830164802725884898839354, Error: Python int too large to convert to C long
```

**Comment by mayfield** [comment]
I'm not really sure what the best course of action here is.  I can envision use-cases where a customer knows their data fits into fast/small types like `numpy.int64`.  Even if they defined their database columns loosely (*i.e. scale == 38*), they might want to *select* the faster types for particular query to match the expected dataset.

Likewise, I would not be surprised if some customers would want full control over the column conversion process.  Eg. lazy row conversion, manual row conversion or just a raw mode that doesn't offer any sort of conversion for power users.  For example, one could imagine a processing pipeline that queries a large set of data and does client side filtering of a single column, perhaps a simple string column.  If their filter criteria is met on that one column they may then want to engage in the more expense conversion of the rest of the data in that row, but only then.  For large datasets that have narrow selection criterias the performance benefits to lazy/manual column conversion could be massive.

Some of these scenarios make me think that allowing the customer to specify some of the behavior for row conversion or even having column conversion overrides could offer a variety of targeted performance opportunities.  The snowflake CLI may be one of those clients as its' output language is always text.  It could probably skip decimal conversion for number types in some cases as an example.

In summary, I'm curious if this sort of configurability has ever been considered and/or ruled out.  If you believe there is merit to the discussion I have some more ideas about how we could specifically modify the connector to give the user some control over the column conversion process and also to the type correctness.  Ie. offering a **strict** mode that always uses `decimal.Decimal` for numbers, an **auto_types** mode that evaluates the *scale* and *precision* settings to choose the most appropriate type for a particular column (the current behavior), and a **raw**/**simple** mode that bypasses as much type conversion as it can for users that may want to engage in their own decoding.

**Comment by smtakeda** [comment]
Thanks for reporting issues. I admit ``numpy`` support is fairly basic, and it is just only because of ``snowflake-sqlalchemy`` integration along with ``pandas``/``jupyter``. Mostly interactive use. The precision and scale are not accurately considered at the moment. The output is just expected to be ``numpy`` compatible. I definitely want to fix them if the issues exist.

The converter class became pluggable since 1.3.12, but the interface is not documented or advertised as I'm not fully confident that the design is right.
https://github.com/snowflakedb/snowflake-connector-python/blob/master/connection.py#L77

The main purpose was to split the converter for SnowSQL (Snowflake CLI) from general uses. In fact, its converter doesn't convert the input data for FIXED and REAL cases but pass the text data as is to the SnowSQL:
https://github.com/snowflakedb/snowflake-connector-python/blob/master/converter_snowsql.py

Other DB drivers have some way of overriding converters, for example:
https://docs.python.org/3.6/library/sqlite3.html#sqlite3.register_converter (sqlite3 to Python)
https://docs.python.org/3.6/library/sqlite3.html#sqlite3.register_adapter (Python to sqlite3)

I recall mysql driver has no-converter mode that seems for benchmark.

For customer's control, I would prefer to finalize the design of pluggable converter class so that any data conversion can override all or one of them. Any feedback would be welcome.

**Comment by mayfield** [comment]
That's good info to know.  I've been working on micro optimizations to that very code assuming they were private APIs (_ prefix);  That could be a problem now..  I'll have to stew on this a bit.

**Comment by ericksc** [comment]
I am using snowflake-connector-python 1.8.5  

I am getting float loss precision. I tried to insert a number like 0.12345678901 at real column but a received 0.123456789.

is it a known issues?

!!! error : assert 0.123456789 == 0.12345678901
  -0.123456789
  +0.12345678901

**Comment by snowstakeda** [comment]
@ericksc yes, it is known issue. From Python connector 2.1.0, when the arrow format is enabled globally (~ Dec 8th), it will be addressed for fetching data but binding and update still have the precision loss.
